diffsims/simulations/simulation2d.py

```diff
@@ -0,0 +1,619 @@
+            the center of the image.
+        in_plane_angle: float, optional
+            In plane rotation of the pattern in degrees
+        mirrored: bool, optional
+            Whether the pattern should be flipped over the x-axis,
+            corresponding to the inverted orientation
+
+        Returns
+        -------
+        diffraction-pattern : numpy.array
+            The simulated electron diffraction pattern, normalised.
+
+        Notes
+        -----
+        If don't know the exact calibration of your diffraction signal using 1e-2
+        produces reasonably good patterns when the lattice parameters are on
+        the order of 0.5nm and the default size and sigma are used.
+        """
+        if direct_beam_position is None:
+            direct_beam_position = (shape[1] // 2, shape[0] // 2)
+        transformed = self._get_transformed_coordinates(
+            in_plane_angle,
+            direct_beam_position,
+            mirrored,
+            units="pixel",
+            calibration=calibration,
+        )
+        in_frame = (
+            (transformed.data[:, 0] >= 0)
+            & (transformed.data[:, 0] < shape[1])
+            & (transformed.data[:, 1] >= 0)
+            & (transformed.data[:, 1] < shape[0])
+        )
+        spot_coords = transformed.data[in_frame].astype(int)
+
+        spot_intens = transformed.intensity[in_frame]
+        pattern = np.zeros(shape)
+        # checks that we have some spots
+        if spot_intens.shape[0] == 0:
+            return pattern
+        else:
+            pattern[spot_coords[:, 0], spot_coords[:, 1]] = spot_intens
+            pattern = add_shot_and_point_spread(pattern.T, sigma, shot_noise=False)
+        return np.divide(pattern, np.max(pattern))
+
+    @property
+    def num_phases(self):
+        """Returns the number of phases in the simulation"""
+        if hasattr(self.phases, "__len__"):
+            return len(self.phases)
+        else:
+            return 1
+
+    @property
+    def has_multiple_phases(self):
+        """Returns True if the simulation has multiple phases"""
+        return self.num_phases > 1
+
+    @property
+    def has_multiple_rotations(self):
+        """Returns True if the simulation has multiple rotations"""
+        if isinstance(self.rotations, Rotation):
+            return self.rotations.size > 1
+        else:
+            return [r.size > 1 for r in self.rotations]
+
+    def get_current_coordinates(self):
+        """Returns the coordinates of the current phase and rotation"""
+        if self.has_multiple_phases:
+            return copy.deepcopy(
+                self.coordinates[self.phase_index][self.rotation_index]
+            )
+        elif not self.has_multiple_phases and self.has_multiple_rotations:
+            return copy.deepcopy(self.coordinates[self.rotation_index])
+        else:
+            return copy.deepcopy(self.coordinates)
+
+    def get_current_rotation(self):
```

Comment:
Bit of a misleading function name, since the rotations are normally `Rotation` objects, but the docstring clarifies.

Reply:
Done!